examples/text/plot_document_clustering.py

```diff
@@ -3,284 +3,344 @@
+# :class:`~sklearn.feature_extraction.text.TfidfTransformer` instance. In this
+# case we also add LSA to the pipeline to evaluate the `"k-means++"` init.
+
+from sklearn.feature_extraction.text import HashingVectorizer
+from sklearn.feature_extraction.text import TfidfTransformer
+
+hasher = HashingVectorizer(
+    stop_words="english",
+    alternate_sign=False,
+    norm=None,
+)
+make_pipeline(TruncatedSVD(n_components=100), Normalizer(copy=False))
+
+lsa_vectorizer = make_pipeline(hasher, TfidfTransformer(), lsa)
 
-print("Clustering sparse data with %s" % km)
 t0 = time()
-km.fit(X)
-print("done in %0.3fs" % (time() - t0))
-print()
+X = lsa_vectorizer.fit_transform(dataset.data)
+print(f"vectorizing done in {time() - t0:.3f} s")
 
+fit_and_evaluate(kmeans, X, name="KMeans with\nhashing and LSA")
 
 # %%
-# Performance metrics
-# -------------------
-
-print("Homogeneity: %0.3f" % metrics.homogeneity_score(labels, km.labels_))
-print("Completeness: %0.3f" % metrics.completeness_score(labels, km.labels_))
-print("V-measure: %0.3f" % metrics.v_measure_score(labels, km.labels_))
-print("Adjusted Rand-Index: %.3f" % metrics.adjusted_rand_score(labels, km.labels_))
-print(
-    "Silhouette Coefficient: %0.3f"
-    % metrics.silhouette_score(X, km.labels_, sample_size=1000)
-)
+# Plot clustering evaluation metrics
+# ==================================
 
-print()
+import pandas as pd
+import matplotlib.pyplot as plt
 
+df = pd.DataFrame(scores).set_index("estimator")
+
+ax = df.plot.barh()
+ax.set_ylabel("")
+ax.set_xlabel("Clustering scores")
+plt.tight_layout()
+
+# %%
+# The Silhouette Coefficient improves when using LSA. It suffers from the
+# phenomenon called "Concentration of Measure" or "Curse of Dimensionality" for
+# high dimensional datasets such as text data. Other measures such as the
+# V-measure and the Adjusted Rand Index are information-theory-based evaluation
+# scores: they are not affected by the curse of dimensionality as they are only
+# based on cluster assignments rather than distances.
 
 # %%
+# Plot clustering time
+# ====================
 
-if not opts.use_hashing:
-    print("Top terms per cluster:")
-
-    if opts.n_components:
-        original_space_centroids = svd.inverse_transform(km.cluster_centers_)
-        order_centroids = original_space_centroids.argsort()[:, ::-1]
-    else:
-        order_centroids = km.cluster_centers_.argsort()[:, ::-1]
-
-    terms = vectorizer.get_feature_names_out()
-    for i in range(true_k):
-        print("Cluster %d:" % i, end="")
-        for ind in order_centroids[i, :10]:
-            print(" %s" % terms[ind], end="")
-        print()
+df = pd.DataFrame(train_times).set_index("estimator")
+ax = df.plot.barh()
+ax.set_xlabel("Clustering times (s)")
+ax.set_ylabel("")
+plt.tight_layout()
+
+# %%
+# :class:`~sklearn.cluster.MiniBatchKMeans` requires less time since it runs
+# computations on batches of data. This comes at the expense of clustering
+# quality.
```

Comment:
We should also discuss the impacts of LSA vs raw tf-idf vectors, hashing vs no-hashing and add a concluding section with global recommendations that take both computation aspects and evaluation metrics into account: using `k-means++` on LSA reduced data is both good in terms of metrics, stability and computing speed.

Reply:
For the stability part, we could have a dedicated section that shows that running kmeans on raw tf-idf vectors with a single `kmeans++` init can often lead to degenerate (highly imbalanced) clustering results.